Vizzuality/gfw
# Feature/static map and dashboard pages

## Overview

Updates the map and dashboard routes to use next v9's `getStaticPaths` and `getStaticProps` when building the app. With this comes a few positive side effects:
1. we are now using the new nextjs [catch all](https://nextjs.org/docs/routing/dynamic-routes#catch-all-routes) which allows for a path of `/map/` to exist without needing a rewrite on the server to `/map/global/`. This solves the issues we were having with query params not updating and is more inline with what we are trying to achieve.
2. we can pre-build the country pages so that they are served statically, making them lighter and faster
3. we can now use the [next-sitemap](https://github.com/iamvishnusankar/next-sitemap) library to parse our statically built pages and dynamically output a sitemap to our public folder. This has some cool features:
- we can trigger out sitemap with the npm `postbuild` script allowing us to rebuild the sitemap whenever we deploy.
- the sitemap will always be up to date with the current page routes of the app
- a `robots.txt` files is output in parallel with the sitemap referencing any (multiple) sitemaps generated and with a size limit. This helps us to follow SEO best practices. 

## Notes

The `next-sitemap` library has the following config. Note the excluded paths and transform functions for handling exceptions to the rule.

```
module.exports = {
  siteUrl: 'https://www.globalforestwatch.org',
  generateRobotsTxt: true,
  sitemapSize: 5000,
  changefreq: 'weekly',
  exclude: [
    '/404',
    '/search',
    '/thank-you',
    '/browser-support',
    '/terms',
    '/my-gfw',
    '/privacy-policy',
    '/subscribe',
  ],
  transform: (config, url) => {
    if (url.includes('/embed')) {
      return null;
    }

    if (url.includes('/map') || url.includes('/dashboards')) {
      return {
        loc: url,
        changefreq: config.changefreq,
        priority: '0.7',
        lastmod: config.autoLastmod ? new Date().toISOString() : undefined,
      };
    }

    return {
      loc: url,
      changefreq: config.changefreq,
      priority: '1.0',
      lastmod: config.autoLastmod ? new Date().toISOString() : undefined,
    };
  },
};
```

## Testing

- check out the `/sitemap.xml`
- visit the global map `/map` and make sure query params are updating
- visit a country map page through click interaction
- visit the dashboards and change countries
- visit a widget embed page
